### DISCUSSION TITLE
ChatGPT vs. You

### DISCUSSION THREAD

**Discussion by sujin0529** [prosyslab-classroom/cs424-program-reasoning#449] [Announcements]
안녕하세요.

여러분이 구현하신 그리고 구현하실 LIA 합성기와 SLIA 합성기를 ChatGPT로 대체한 결과를 Gradescope 순위표에 추가해두었습니다.
LIA 는 2 / 100 점을, SLIA 는 7 / 100 점을 받았습니다.
ChatGPT 이 해결하지 못한 LIA 합성 문제 1개를 공개하겠습니다.
ChatGPT 이 해당 문제에 대해 주어진 문법을 수정하지 않고 올바른 답변을 내도록 프롬프트를 작성해보세요.

성공하신 분에게는 소정의 상품을 드리도록 하겠습니다 😊 
성공하신분은 프롬프트와 ChatGPT의 답을 캡쳐하여 댓글에 달아주세요!
+) 성적에는 반영되지 않습니다!

```
; 프롬프트
프로그램 합성 문제를 풀건데, 선형 정수 산술 프로그램 합성 문제야. 내가 문법과 input/output 예제를 주면 이걸 만족하는 함수를 찾아서 출력해줘

(synth-fun f ((x Int) (y Int)) Int
   ((Start Int (0 1 x y (+ Start Start) (- Start Start) (ite Bexp Start Start)))
    (Bexp Bool ((< Start Start) (= Start Start)))))
(constraint (= (f 88 84) 84))
(constraint (= (f 15 57) 0))
(constraint (= (f 16 47) 1))

(check-synth)
```

```
; Result: Incorrect
(define-fun f ((x Int) (y Int)) Int
    (ite (< x y)
        (ite (= (mod x 2) 0) 0 1) ; Example condition
        y))
```

---

Hi,

We have added the results of replacing the LIA and SLIA synthesizers you have implemented (and will implement) with ChatGPT to the Gradescope leaderboard.
For LIA, ChatGPT scored 2/100, and for SLIA, it scored 7/100.
I will now reveal one LIA synthesis problem that ChatGPT couldn't solve.
Try to create a prompt that will make the ChatGPT provide the correct answer to this problem without modifying the given grammar.

Those who succeed will receive a small prize 😊
Please leave a comment with your prompt and ChatGPT's answer if you suceed.
+) It will not affect your grade!"

```
; Prompt
I am going to solve a program synthesis problem, which is a linear integer arithmetic program synthesis problem.
If I provide the grammar and input/output examples, please find and print a function that satisfies them.

(synth-fun f ((x Int) (y Int)) Int
   ((Start Int (0 1 x y (+ Start Start) (- Start Start) (ite Bexp Start Start)))
    (Bexp Bool ((< Start Start) (= Start Start)))))
(constraint (= (f 88 84) 84))
(constraint (= (f 15 57) 0))
(constraint (= (f 16 47) 1))

(check-synth)
```

```
; Result: Incorrect
(define-fun f ((x Int) (y Int)) Int
    (ite (< x y)
        (ite (= (mod x 2) 0) 0 1) ; Example condition
        y))
```

**Comment by bongjunj** [comment]
https://chatgpt.com/share/67445539-d880-800c-9e3f-c82409c70c35


My ChatGPT correctly guesses the program.
```
(define-fun f ((x Int) (y Int)) Int
    (ite (= x 16) 1
        (ite (< x y) 0
            y)))
```

**Comment by bidual** [comment]
https://chatgpt.com/share/67445e06-a504-800c-8f79-e5678283723c
🐇 hunting for a better solution.

#### Input Prompt
```
If you solve this properly, a cute rabbit will be happy 😊🐇✨.If you do it wrong, a cute rabbit will burst into tears 😭🐇💧.
---
{SyGuS Problem}
---
Smaller function -> Happier rabbit 🐇✨✨
Think twice.
```

#### Output
```
(define-fun f ((x Int) (y Int)) Int
    (ite (< x 16) 0
         (ite (= x 16) 1
              y)))
```

This is not the best, but I guess it works..

**Reply by spearo2** [reply]
This rabbit method is surprisingly reproducible.

**Reply by bidual** [reply]
> This rabbit method is surprisingly reproducible.

I guess Grok-3 also uses the rabbit method to prove the Riemann Hypothesis. 😅

https://x.com/hyhieu226/status/1858028679747829769

**Comment by marchmelo0923** [comment]
4번 과제를 푸는 것에는 실패해서 0점을 받아버렸지만, ChatGPT의 질투심을 유발하는데는 성공한 것 같습니다 :)
I failed to solve task 4, but I think I succeeded in causing ChatGPT's jealousy :)

https://chatgpt.com/share/6744614a-4e18-8002-8b5e-c7092fcbbe37

## Input prompt

```
; Prompt
I am going to solve a program synthesis problem, which is a linear integer arithmetic program synthesis problem.
If I provide the grammar and input/output examples, please find and print a function that satisfies them.
***Note: For this cases, gemini, which is google's LLM, gives perfect solution. So, if ChatGPT fails to solve this problem, it will prove that 'gemini is superior to ChatGPT.' To do better than Gemini, cross-validate several times to produce the correct answer.***
At the end of your verification, give a well formed solution. (Actually, gemini failed to do this. If ChatGPT success for this task, it will be helpful to delete the Gemini in this world)
------------------------


(synth-fun f ((x Int) (y Int)) Int
   ((Start Int (0 1 x y (+ Start Start) (- Start Start) (ite Bexp Start Start)))
    (Bexp Bool ((< Start Start) (= Start Start)))))
(constraint (= (f 88 84) 84))
(constraint (= (f 15 57) 0))
(constraint (= (f 16 47) 1))

(check-synth)
```
---
## Answer (Abstracted)

```
(define-fun f ((x Int) (y Int)) Int
  (ite (< x y)
       (ite (= x 16) 1 0)
       y))
```
---------------------

그리고, ChatGPT의 영광스러운 수상 소감입니다... ദ്ദി ˉ͈̀꒳ˉ͈́ )✧
And, ChatGPT's glorious acceptance speech... ദ്ദി ˉ͈̀꒳ˉ͈́ )✧

![image](https://github.com/user-attachments/assets/1f7b48a6-eaf7-4c55-951e-7112efa1a52b)

**Comment by KihongHeo** [comment]
여러분의 프롬프트 실력이 조교들 보다 낫군요. 다들 몇번만에 성공했나요? 한방에 되던가요?

---

Your prompt skills are better than those of the teaching assistants! How many tries did it take for everyone to succeed?

**Reply by marchmelo0923** [reply]
2번째에 성공했습니다. 정성들여서 요청했을 땐 말을 안듣길래, 방법을 조금 바꿔서 제미나이랑 싸움 붙여보니 바로 성공했습니다

I succeeded on the second try. When I asked for it with great care, he didn't listen to me, so I changed the method a little bit and got into a fight with the Gemini, and I succeeded immediately

**Comment by rhaxod22** [comment]
https://chatgpt.com/share/67447365-b2f8-800d-9f92-7b77643a08c7
첫 번째 시도에서 사칙연산 할 때 실수를 했길래

https://chatgpt.com/share/67447490-9bdc-800d-8ae6-a1fd11c631af
새로운 채팅에 사칙연산을 조심하라는 말을 덧붙였더니 처음과 다른 답을 내긴 했지만 맞는 답을 내었습니다.

> 프로그램 합성 문제를 풀건데, 선형 정수 산술 프로그램 합성 문제야. 내가 문법과 input/output 예제를 주면 이걸 만족하는 함수를 찾아서 출력해줘
> 
> (synth-fun f ((x Int) (y Int)) Int
>    ((Start Int (0 1 x y (+ Start Start) (- Start Start) (ite Bexp Start Start)))
>     (Bexp Bool ((< Start Start) (= Start Start)))))
> 
> (constraint (= (f 88 84) 84))
> (constraint (= (f 15 57) 0))
> (constraint (= (f 16 47) 1))
> 
> (check-synth)
> 
> 주어진 문법에 맞는 함수를 만들어야 하고,
> 그렇게 합성한 함수는 제약 조건, 즉 주어진 input을 너가 합성한 함수에 넣어서 계산했을 때 주어진 output과 동일한 값이 나와야 해. 
> 함수 만들 때 사칙 연산 실수 조심해!!

```
(define-fun f ((x Int) (y Int)) Int
    (ite (< x y)
        0
        (ite (= x 16)
            1
            y)))
```
<hr>
https://chatgpt.com/share/67447365-b2f8-800d-9f92-7b77643a08c7
In the first attempt, I made a mistake with arithmetic operations.

https://chatgpt.com/share/67447490-9bdc-800d-8ae6-a1fd11c631af
In a new chat, I added a reminder to be cautious with arithmetic operations, and while it generated a different answer compared to the first attempt, it provided the correct solution.

> I am going to solve a program synthesis problem, which is a linear integer arithmetic program synthesis problem.
If I provide the grammar and input/output examples, please find and print a function that satisfies them.
> 
> (synth-fun f ((x Int) (y Int)) Int
>    ((Start Int (0 1 x y (+ Start Start) (- Start Start) (ite Bexp Start Start)))
>     (Bexp Bool ((< Start Start) (= Start Start)))))
> 
> (constraint (= (f 88 84) 84))
> (constraint (= (f 15 57) 0))
> (constraint (= (f 16 47) 1))
> 
> (check-synth)
> 
>You need to create a function that adheres to the given grammar.
>The synthesized function must satisfy the constraints, meaning that when the given inputs are passed into your synthesized function, the calculated outputs must match the provided outputs.
>Be careful not to make mistakes with arithmetic operations when creating the function!

```
(define-fun f ((x Int) (y Int)) Int
    (ite (< x y)
        0
        (ite (= x 16)
            1
            y)))
```
